pkg/streamingpromql/operators/aggregations/limitklimitratio/operator.go

```diff
@@ -0,0 +1,409 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package limitklimitratio
+
+import (
+	"context"
+	"slices"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/operators/aggregations"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+// Note - this ratio sampler is used for limit_ratio to determine a hash for a series labels and
+// to use this hash for the ratio comparison.
+// In its current use, this RatioSampler is called via AddRatioSample(), which passes in the labels for
+// each point comparison. This results in the hash for the series being redundantly calculated.
+// A future enhancement will be to modify this utility to re-use a hash between ratio comparisons.
+var ratiosampler promql.RatioSampler = promql.NewHashRatioSampler()
+
+// discardPointAtStepK is the function we use in limitk mode to determine when to stop accumulating points for a series in a group.
+// Once the number of series at the given step reaches k no more series in this group are considered.
+// Note that k may not be step invariant, so we accumulate into a collection of sparse series for each group.
+// ie if k is [ 1, 2, 3 ] we would expect to see the series within a group such as;
+// {series1}[v, v, v]
+// {series2}[_, v, v]
+// {series3}[_, _, v]
+var discardPointAtStepK = func(_ types.SeriesMetadata, step int, g *limitGroup, p stepArgument) bool {
+	k := p.(*kParam).k[step]
+	return int64(g.countAtStep(step)) == k
+}
+
+// discardPointAtStepR is the function we use in limit_ratio mode to determine if a series should be accumulated for a group at the given step.
+// As above, the ratio value for each step may not be step invariant so each series in the group may be sparse.
+var discardPointAtStepR = func(metadata types.SeriesMetadata, step int, g *limitGroup, p stepArgument) bool {
+	r := p.(*ratioParam).r[step]
+	return !ratiosampler.AddRatioSample(r, &promql.Sample{Metric: metadata.Labels})
+}
```

Comment:
I'm wondering if we can rejig the design slightly to avoid the casts, and make the logic a little easier to follow - there are a lot of moving parts here.

What if we made `limitGroup` an interface, with two implementations (one for `limitk`, and another for `limit_ratio`)? When we create the group, we could pass the appropriate parameter type, and `limitGroup` could have a `shouldDiscard` method.

The logic shared between both `limitk` and `limit_ratio` groups could be solved by embedding a common type in both group implementations.